book_source/03_topical_pages/09_standalone_tools.Rmd

```diff
@@ -177,13 +177,26 @@ earthengine authenticate
 
+Whenever a data product is requested the output files are stored in the inputs table of BETYdb. Subsequently when the same product is requested again with a different date range but with the same qc, scale, projection the previous file in the db would be extended. The DB would always contain only one file of the same type.
+As an example, if a file `Reykajvik_gee_s2_10.0_NA_1.0_200802174519.nc` containing Sentinel 2 bands for start date: 2018-01-01, end date: 2018-06-30 exists in the DB and the same product is requested again for a different date range one of the following cases would happen,
+
+1. New dates are ahead of the existing file: For example, if the requested dates are start: 2018-10-01, end: 2018-12-31 in this case the previous file will be extended forward meaning the effective start date of the file to be downloaded would be the day after the end date of the previous file record, i.e. 2018-07-01. The new and the previous file would be merged and the DB would now be having data for 2018-01-01 to 2018-12-31.
+
+2. New dates are preceding of the existing file: For example, if the requested dates are start: 2017-01-01, end: 2017-06-30 in this case the effective end date of the new download would be the day before the start date of the existing file, i.e., 2017-12-31. The new and the previous file would be merged and the file in the DB would now be having data for 2017-01-01 to 2018-06-30. 
+
+3. New dates contain the date range of the existing file: For example, if the requested dates are start: 2016-01-01, end: 2019-06-30 here the existing file would be replaced entirely with the new file. A more efficient way of doing this could be to divide your request into two parts, i.e, first request for 2016-01-01 to 2018-01-01 and then for 2018-06-30 to 2019-06-30.
+
+When a processed data product such as SNAP-LAI is requested, the raw product (here Sentinel 2 bands) used to create it would also be stored in the DB. If the raw product required for creating the processed product already exists for the requested time period, the processed product would be created for the entire time period of the raw file. For example, if Sentinel 2 bands are present in the DB for 2017-01-01 to 2017-12-31 and SNAP-LAI is requested for 2017-03-01 to 2017-07-31, the output file would be containing LAI for 2017-01-01 to 2017-12-31.
 
 **Input data**: The input data must be the coordinates of the area of interest (point or polygon type) and the name of the site/AOI. These information must be provided in a GeoJSON file.
 
 **Output data**: The output data returned by the GEE are in the form of netCDF files. When using AppEEARS, output is in the form of netCDF files if the AOI type is a polygon and in the form of csv files if the AOI type is a point.
 
+The output files are created with the following naming convention: sitename_source_collection_scale_projection_qc_TimeStampOfFileCreation
```

Comment:
I don't have siteid here because I think that info is redundant and currently, siteid never goes beyond `call_remote_process` and reach the Python functions. But if it's required to make consistent with PEcAn guidelines I will do so.

Reply:
-Yeah I think we would like to have siteid instead of sitename because it is less ambiguous (also some sites really have long names with spaces in them (e.g. `Ontario - Turkey Point Mature Deciduous (CA-TPD)`) whereas siteid always have an expected form, e.g. `1000000109` turns into `site_1-109` in the filename. 

-Do we really need the time stamp in the filename? 

-Also we can think about the difference between foldername and filename, for example, foldername could be `source_siteid` and the filename can be `collection_scale_projection_qc_siteid`. 
```
GEE_site_1-109
  \_  S2_10-1_site_1-109.nc
  \_  L8_10-1_site_1-109.nc
```
We can play around with these a little

Reply:
>  e.g. 1000000109 turns into site_1-109 in the filename.

Sorry not aware of the logic. How does this happen?

> -Do we really need the time stamp in the filename?

We need something to make the file names unique otherwise when the same data is requested again in the same outdir then the previous file will be directly overwritten before we can do anything with it. I have also tried a different approach - when the merge flag is true, I renamed the original final with some random string before starting the new download and then after merging was completed I changed it back to its original name. This worked fine but I think it's risky because if anything unexpectedly goes wrong during download or merging then the file name in the DB won't be updated and would be pointing towards an incorrect file. I used timestamp because by nature it would always be unique and human-readable but I can use the cryptography functions to generate random strings or anything else as well to make every file name unique.

Reply:
>Sorry not aware of the logic. How does this happen?

Sorry I was being cryptical, once we know about the `siteid`, we create this string, like `paste0(siteid %/% 1e+09, "-", siteid %% 1e+09)` You can see an example [here](https://github.com/PecanProject/pecan/blob/develop/modules/data.land/R/ic_process.R#L62-L65).

>when the same data is requested again in the same outdir then the previous file will be directly overwritten before we can do anything with it. 
Hmm this doesn't sound quite right, before we do anything we query the DB to see if there has been any input file for this site, previous file should be returned from that query with date info as they are stored in the DB. Sorry, I still haven't been able to check you code in detail yet, I don't know how the merge flag works. I'll come back to this discussion once I work through your PR